File: go/dp_two_dim/518_change.go
package dptwodim

// change 零钱兑换II
// 计算并返回可以凑成总金额的硬币组合数 （硬币数量无限个）
// 完全背包视角 从新思考这个问题
func change(amount int, coins []int) int {
	// coins[i] 表示物品的重量  和 价值
	// dp[j] 表示容量为j 的背包中，能够填满的方法有dp[j]种
	// dp[j] += dp[j - conis[i]]

	dp := make([]int, amount+1)
	dp[0] = 1

	for i := 0; i < len(coins); i++ {
		for j := coins[i]; j < amount+1; j++ {
			dp[j] += dp[j-coins[i]]
		}
	}

	return dp[amount]
}

File: rust/src/dp_two_dim/_518_change.rs
/// change 零钱兑换II
/// 计算并返回可以凑成总金额的硬币组合数
/// 完全背包视角 重新解读此题
pub fn change(amount: i32, coins: Vec<i32>) -> i32 {
    // dp[j] 表示填满容量为j 的背包，有dp[j] 中方法
    // dp[j] += dp[j-coins[i]]

    let amount = amount as usize;
    // 定义dp
    let mut dp = vec![0; amount + 1];
    // 初始化dp
    dp[0] = 1;

    for i in 0..coins.len() {
        for j in coins[i] as usize..=amount {
            dp[j] += dp[j - coins[i] as usize];
        }
    }

    dp[amount]
}
